File: Python/프로그래머스/Level2/[1차] 뉴스 클러스터링.py
import re
str1 = 'FRANCE'
str2 = 'french'


def solution(str1, str2):
    answer = 0
    s1, s2, news1, news2 = ([], [], [], [])
    for i in range(len(str1)-1):
        s1.append(str1[i:i+2].lower())
    for i in range(len(str2)-1):
        s2.append(str2[i:i+2].lower())
    # print(f"s1:{s1} s2:{s2}")

    for s in s1:  # 특수문자, 숫자제외
        if len(re.findall('[a-zA-Z]', s)) == 2:
            news1.append(s)
    for s in s2:
        if len(re.findall('[a-zA-Z]', s)) == 2:
            news2.append(s)
    # print(f"news1:{news1} news2:{news2}")

    union, intersection = 0, 0
    for el in set(news1+news2):
        union += max(news1.count(el), news2.count(el))
        intersection += min(news1.count(el), news2.count(el))
    # print(f"합: {union} 교: {intersection}")
    if union == 0:  # 분모가 0일 때 처리
        return 65536
    answer = int((intersection / union)*65536)
    print(answer)
    return answer


solution(str1, str2)

File: Javascript/프로그래머스/Level2/[1차] 뉴스 클러스터링.js
function solution(str1, str2) {
    var answer = 0;
    let ar1 = [];
    let ar2 = [];
    //소문자 변환
    str1 = str1.toLowerCase();
    str2 = str2.toLowerCase();

    //2글자씩 띄어주기
    for (let i = 0; i < str1.length - 1; i++) {
        let tmp = str1.slice(i, i + 2).match(/[a-z]+/g);
        // console.log(tmp);
        if (tmp !== null && tmp[0].length > 1) {
            ar1.push(tmp[0]);
        }
    }

    for (let i = 0; i < str2.length - 1; i++) {
        let tmp = str2.slice(i, i + 2).match(/[a-z]+/g);
        if (tmp !== null && tmp[0].length > 1) {
            ar2.push(tmp[0]);
        }
    }
    // console.log(ar1)
    // console.log(ar2)
    let set = new Set([...ar1, ...ar2]); // 모든요소
    let common = 0;
    let union = 0;

    set.forEach((el) => {
        let cnt1 = ar1.filter((each) => each === el).length;
        let cnt2 = ar2.filter((each) => each === el).length;
        common += Math.min(cnt1, cnt2);
        union += Math.max(cnt1, cnt2);
    });
    // console.log(common)
    // console.log(union)
    if (union !== 0) {
        return Math.floor((common / union) * 65536);
    } else {
        return 65536;
    }
}
